src/test/java/org/tarantool/ClientReconnectClusterIT.java

```diff
@@ -57,61 +64,271 @@ public static void tearDownEnv() {
+     * 2. shutdown srv1 - fail (client will not reconnect to srv2 because latest is out of the list)
+     */
+    @Test
+    @DisplayName("applied new nodes and stayed connected to the current node")
+    void testUpdateNarrowNodeList() {
+
+        String service1Address = "localhost:" + PORTS[0];
+        String service2Address = "127.0.0.1:" + PORTS[1];
+
+        String infoFunctionName = "getAddresses";
+        String infoFunctionScript = makeDiscoveryFunction(infoFunctionName, Arrays.asList(service1Address));
+
+        control.openConsole(SRV1).exec(infoFunctionScript);
+
+        final TarantoolClusterClient client = makeClientWithDiscoveryFeature(
+                service1Address,
+                infoFunctionName,
+                service1Address,
+                service2Address
+        );
+
+        int[] ids = makeAndFillTestSpace(client, "rr_test3");
+        final int spaceId = ids[0];
+        final int pkId = ids[1];
+
+        expectConnected(client, spaceId, pkId);
+
+        stopInstancesAndAwait(SRV1);
+        expectDisconnected(client, spaceId, pkId);
+
+        stopInstancesAndAwait(SRV2);
+        stopInstancesAndAwait(SRV3);
+    }
+
+    /**
+     * Before fetch <code>client = { srv1, srv2, srv3 }</code>
+     * After fetch <code>client = { srv1, srv2, srv3 }</code>
+     *
+     * 1. fetch nodes - ok (client will ignore empty list)
+     * 2. shutdown srv1 - ok
+     * 3. shutdown srv2 - ok
+     * 4. shutdown srv3 - fail
+     */
+    @Test
+    @DisplayName("applied empty list and stayed connected to the current node")
+    void testUpdateEmptyNodeList() {
+
+        String service1Address = "localhost:" + PORTS[0];
+        String service2Address = "127.0.0.1:" + PORTS[1];
+        String service3Address = "localhost:" + PORTS[2];
+
+        String infoFunctionName = "getAddresses";
+        String infoFunctionScript = makeDiscoveryFunction(infoFunctionName, Collections.emptyList());
+
+        control.openConsole(SRV1).exec(infoFunctionScript);
+
+        final TarantoolClusterClient client = makeClientWithDiscoveryFeature(
+                service1Address,
+                infoFunctionName,
+                service1Address,
+                service2Address,
+                service3Address
+        );
 
+        int[] ids = makeAndFillTestSpace(client, "rr_test4");
+        final int spaceId = ids[0];
+        final int pkId = ids[1];
+
+        expectConnected(client, spaceId, pkId);
+
+        stopInstancesAndAwait(SRV1);
+        expectConnected(client, spaceId, pkId);
+
+        control.stop(SRV2);
+        expectConnected(client, spaceId, pkId);
+
+        control.stop(SRV3);
+        expectDisconnected(client, spaceId, pkId);
+    }
+
+    /**
+     * Before fetch <code>client = { srv1, srv2, srv3 }</code>
+     * After fetch <code>client = { srv1, srv2, srv3 }</code>
+     *
+     * 1. fetch with an exception - ok (client will ignore the failure)
```

Comment:
Some extra tests were added to ClusterServiceStoredFunctionDiscovererIT and ClientReconnectClusterIT to check this cases. 
* `error('msg')` will cause the TarantoolException (same as an invocation undefined function)
* `42` and any other data being different than String will be filtered out (as a result we'll get an empty list instead of exception)

Reply:
I mean `return 42`, not `return {42}`. But filtering of non-string data from a list of nodes does not seems to be good behaviour for me. It is better to log an error and discard the entire list (as well as for `error('msg')` case). I think `return 42` case also should lead to the same behaviour.

Reply:
1. not really, if at least one object from the list isn't a string - the whole list will be discarded and an empty list will be returned. In this case, we can throw an exception instead of returning an empty list.
2. Indeed, `ClusterServiceStoredFunctionDiscovererIT.testWrongResultData` test generates `function getAddresses() return 42 end`, not `{42}` and the client still parses it as an array with one element equal 42. And this fact forces me to check a result to be sure we received the strings. Again we can terminate the operation with an exception mentioned above.

Reply:
1. I misread your answer about filtering. The whole result will be filtered out in any validation error, correct? The behaviour in all such cases now is to set empty list of nodes (and so fail all future requests). I think that when a background cluster discovery job fails due to any reason we should ignore that result (don't replace a list of nodes with empty one, don't trigger reconnection — nothing like that), warn a user about that via logging and schedule the next discovery task as usual.
2. A function in Lua can return several results: `function f() return 1, 2, 3 end` (it is often called multireturn). So result of a call is always a list even if we receive a single value. When we return a list, a client will see a singleton list with the returned list as the item. Despite multireturn feature I want a discovery function have the contract with explicit list due to two reasons:

* I don't sure about how much values can be returned with multireturn and prefer to use it only when I know exact count of values at 'compile time' and only for specific tasks like reporting failure (`return nil, err_msg`) or extending existing format in non-breaking way.
* It will be hard to extend such contract, but if we're use a list we can just add one more return value with, say, some metainformation.

BTW, it worth to ignore 2nd and following return values in order to be able to extend the contract in a way that will not break old clients. This weakness of validation is not obvious, so need to be commented properly in the code.